File: packages/flint-js/src/vegalite/templates/bullet.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import { ChartTemplateDef } from '../../core/types';

/**
 * Bullet chart — a compact KPI panel: one row per label, each showing a measure
 * bar compared against a target marker.
 *
 * Channels:
 *   - y      label (the banded category axis — one row per KPI)
 *   - x      measured value (the bar, drawn from a zero baseline)
 *   - goal   target / comparative value (drawn as a tick across the bar)
 *   - color  optional grouping for the value bar
 *
 * The measure bar reads as length-from-zero (markCognitiveChannel 'length'), so
 * the x scale keeps its zero baseline. The target tick is layered on top and
 * sized a little taller than the bar so it stands out as a reference marker.
 * Qualitative range bands (poor/ok/good) are intentionally left out of this
 * first version; the value-vs-target comparison is the core of the encoding.
 */
export const bulletChartDef: ChartTemplateDef = {
    chart: "Bullet Chart",
    template: {
        encoding: {},
        layer: [
            { mark: { type: "bar", color: "#3a6ea5", height: { band: 0.5 } }, encoding: {} },
            { mark: { type: "tick", color: "#1a1a1a", thickness: 3, opacity: 1 }, encoding: {} },
        ],
    },
    channels: ["y", "x", "goal", "color", "column", "row"],
    markCognitiveChannel: 'length',
    declareLayoutMode: () => ({
        axisFlags: { y: { banded: true } },
    }),
    instantiate: (spec, ctx) => {
        const { x, y, goal, color, column, row } = ctx.resolvedEncodings;

        if (!spec.encoding) spec.encoding = {};

        if (y) {
            spec.encoding.y = { ...y };
            spec.encoding.y.axis = { ...(spec.encoding.y.axis ?? {}), title: null };
        }
        if (column) spec.encoding.column = column;
        if (row) spec.encoding.row = row;

        // Value bar — length from zero.
        let valueTitle: any;
        if (x) {
            spec.layer[0].encoding.x = { ...x };
            spec.layer[0].encoding.x.scale = { ...(x.scale ?? {}), zero: true };
            valueTitle = x.title ?? x.field;
            spec.layer[0].encoding.x.axis = { ...(x.axis ?? {}), title: valueTitle };
        }
        if (color) spec.layer[0].encoding.color = color;

        // Target marker — a tick at the goal value, sized a little taller than
        // the bar so it reads as a reference line. It shares the value bar's x
        // scale; give it the same axis title so the shared axis renders once and
        // reads as the measure rather than "value, goal".
        if (goal) {
            spec.layer[1].encoding.x = {
                field: goal.field,
                type: 'quantitative',
                ...(valueTitle != null ? { axis: { title: valueTitle } } : {}),
            };
        }

        // Size the target tick relative to the available band height so it spans
        // a bit more than the value bar regardless of how many rows there are.
        const table = ctx.table ?? [];
        const yField = y?.field;
        const plotHeight = ctx.canvasSize?.height || 300;
        let tickSize = 24;
        if (yField && table.length > 0) {
            const rows = new Set(table.map((r: any) => r[yField])).size || 1;
            tickSize = Math.max(12, Math.min(46, Math.round((plotHeight * 0.62) / rows)));
        }
        spec.layer[1].mark = { ...spec.layer[1].mark, size: tickSize };
    },
};

File: packages/flint-py/flint/vegalite/templates/bullet.py
"""Bullet Chart template — a compact KPI panel.

One row per label, each showing a measure bar compared against its own target,
set against muted gray range bands. Port of vegalite/templates/bullet.ts.
"""
from __future__ import annotations

import json

from ...core import js_round

# Muted grays for the qualitative zones, darkest nearest zero (poorest range).
ZONE_GRAYS = ["#e2e2e2", "#ececec", "#f5f5f5"]
# Goal-attainment colors: muted red for under target, muted green for met target.
STATUS_COLORS = {"below": "#c44e52", "met": "#2f855a"}
STATUS_BELOW = "Below target"
STATUS_MET = "Meets target"


def _is_finite_number(v):
    return isinstance(v, (int, float)) and not isinstance(v, bool) and v == v and v not in (float("inf"), float("-inf"))


def _bullet_declare_layout(cs, table, chart_properties):
    return {"axisFlags": {"y": {"banded": True}}}


def _bullet_instantiate(spec, ctx):
    encs = ctx["resolvedEncodings"]
    x = encs.get("x")
    y = encs.get("y")
    goal = encs.get("goal")
    color = encs.get("color")
    column = encs.get("column")
    row = encs.get("row")

    value_title = (x.get("title") if x.get("title") is not None else x.get("field")) if x else None
    x_axis = {"title": value_title} if value_title is not None else {}

    # Shared category axis at the top level so every layer aligns by row.
    spec["encoding"] = {}
    if y:
        spec["encoding"]["y"] = {**y, "axis": {**(y.get("axis") or {}), "title": None}}
    if column:
        spec["encoding"]["column"] = column
    if row:
        spec["encoding"]["row"] = row

    table = ctx.get("table") or []
    layers = []

    # --- Per-row gray qualitative bands (drawn first, behind everything) ---
    if x and x.get("field") and y and y.get("field") and goal and goal.get("field") and len(table) > 0:
        y_field = y["field"]
        goal_field = goal["field"]
        zone_data = [[], [], []]
        for r in table:
            cat = r.get(y_field)
            g = r.get(goal_field)
            try:
                g = float(g)
            except (TypeError, ValueError):
                continue
            if cat is None or not _is_finite_number(g) or g <= 0:
                continue
            zone_data[0].append({y_field: cat, "__lo": 0, "__hi": 0.25 * g})
            zone_data[1].append({y_field: cat, "__lo": 0.25 * g, "__hi": 0.5 * g})
            zone_data[2].append({y_field: cat, "__lo": 0.5 * g, "__hi": 0.75 * g})
        for i, rows in enumerate(zone_data):
            if len(rows) == 0:
                continue
            layers.append({
                "data": {"values": rows},
                "mark": {"type": "rect", "color": ZONE_GRAYS[i], "opacity": 1},
                "encoding": {
                    "x": {"field": "__lo", "type": "quantitative", "axis": x_axis},
                    "x2": {"field": "__hi"},
                },
            })

    # --- Value bar — length from zero, colored by goal attainment ---
    bar_layer = {"mark": {"type": "bar", "height": {"band": 0.5}}, "encoding": {}}
    if x:
        bar_layer["encoding"]["x"] = {
            **x,
            "scale": {**(x.get("scale") or {}), "zero": True},
            "axis": {**(x.get("axis") or {}), "title": value_title},
        }
    if color:
        # Explicit grouping wins over goal-attainment coloring.
        bar_layer["encoding"]["color"] = color
    elif x and x.get("field") and goal and goal.get("field"):
        bar_layer["transform"] = [{
            "calculate": "datum[{}] >= datum[{}] ? '{}' : '{}'".format(
                json.dumps(x["field"]), json.dumps(goal["field"]), STATUS_MET, STATUS_BELOW),
            "as": "__status",
        }]
        bar_layer["encoding"]["color"] = {
            "field": "__status",
            "type": "nominal",
            "scale": {
                "domain": [STATUS_BELOW, STATUS_MET],
                "range": [STATUS_COLORS["below"], STATUS_COLORS["met"]],
            },
            "legend": {"title": None},
            "title": None,
        }
    layers.append(bar_layer)

    # --- Target marker — a dark tick at the goal, sized to the row band ---
    if goal:
        band = (ctx.get("layout") or {}).get("yStep")
        if band and band > 0:
            tick_size = min(band, max(8, js_round(band * 0.72)))
        else:
            tick_size = 22
        layers.append({
            "mark": {"type": "tick", "color": "#1a1a1a", "thickness": 3, "opacity": 1, "size": tick_size},
            "encoding": {
                "x": {"field": goal["field"], "type": "quantitative", "axis": x_axis},
            },
        })

    spec["layer"] = layers


bullet_chart_def = {
    "chart": "Bullet Chart",
    "template": {"encoding": {}, "layer": []},
    "channels": ["y", "x", "goal", "color", "column", "row"],
    "markCognitiveChannel": "length",
    "declareLayoutMode": _bullet_declare_layout,
    "instantiate": _bullet_instantiate,
}
